Overhaul use of search indexes and add additional filters to the search page

jheretic opened the pull request:
This updates the search postgres search indexes to use trigrams and adds additional search filters. It also fixes the dialog when no search results are found, and applies numerous usability fixes. There's a mostly-complete, more complex implementation of this in the the `search()` method of the preprint model, but that's currently dead code; I'm just leaving it in for the moment before pruning it out in case we run into any problems with this simplified implementation and I need to borrow any code.

To test, make sure you can do partial searches of preprints based on title, handle, author, or abstract (I realize we don't expose author or abstract unless the PDF doesn't exist, but it is searched on the backend), that the 4 sort filters work, and that with a search with no results the sort options and paginations are hidden and there's a message that prompts you to open the add review/request modal.

A note: `sqlite` doesn't support case-insensitive full-text searches, so if you test there then only case-sensitive searches work. This isn't a problem on `postgres`.

harumhelmy reviewed (commented):
The search is working great and omg so so fast! 

The weirdest thing, though, the `SortOptions` component wasn't rendering for me until the one tiny change I made to the conditional logic that decides whether to render it, which I pushed up. Would you mind testing it on your end @jheretic?

jheretic commented:
Ah, good catch. that's a weird thing. It renders both before and after your fix for me. Looking at it, I think the fact that there's a useEffect that watches 'loadingPreprints' and then sets 'loading' based on that might cause a race conditiion? It also looks useless, I think that might be an artifact of adapting the old page to the new hooks. I replaced the one remaining instance of checking 'loading' with 'loadingPreprints' and removed useEffect and it still seems to work fine.

harumhelmy reviewed (approved):
Yessss it's working great! Thanks for identifying that
